### PR TITLE
More explicit time management in BaseCouplingScheme

### PR DESCRIPTION
**Pull request by BenjaminRodenberg**
## Main changes of this PR

* Introduce `_nextTimeWindowSize` as a buffer for time window size received in the participant first method
* Explicitly set `_timeWindowSize = _nextTimeWindowSize` when at end of `advance`
* Use `_computedTimeWindowPart` only to track current point in time, not to move `_timeWindowStartTime` to next window.

## Motivation and additional information

The way how preCICE tracks participant and its internal time is complicated. I'm trying to simplify things with this refactoring.

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [x] ~~I added a changelog file with `make changelog` if there are user-observable changes since the last release.~~ Not relevant, only refactoring
* [x] ~~I added a test to cover the proposed changes in our test suite.~~ Not relevant, only refactoring
* [x] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I sticked to C++17 features.
* [x] I sticked to CMake version 3.16.3.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

**Review by fsimonis — APPROVED** [review]
Looks good and makes sense

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 1]
```diff
@@ -208,7 +209,7 @@ void BaseCouplingScheme::receiveData(const m2n::PtrM2N &m2n, const DataMap &rece
 void BaseCouplingScheme::receiveDataForWindowEnd(const m2n::PtrM2N &m2n, const DataMap &receiveData)
 {
   const double oldComputedTimeWindowPart = _computedTimeWindowPart;
-  _computedTimeWindowPart += getTimeWindowSize(); // such that getTime() in receiveData returns time at end of window
+  _computedTimeWindowPart += _nextTimeWindowSize; // such that getTime() in receiveData returns time at end of window
   this->receiveData(m2n, receiveData);            // receive data for end of window
   _computedTimeWindowPart = oldComputedTimeWindowPart;
```
Comment (uekerman) [inline src/cplscheme/BaseCouplingScheme.cpp:214]: Could we make this more explicit to improve readability?

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 2]
```diff
@@ -281,7 +287,8 @@ void BaseCouplingScheme::initialize(double startTime, int startTimeWindow)
 
   exchangeInitialData();
 
-  _isInitialized = true;
+  _timeWindowSize = _nextTimeWindowSize;
```
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply src/cplscheme/BaseCouplingScheme.cpp]: Took care about this in [d3542d4](https://github.com/precice/precice/pull/1887/commits/d3542d4e7596c2bec954f86cf93d9ca00673eef7). Decided to move it inside `SerialCouplingScheme`.

File: src/cplscheme/BaseCouplingScheme.cpp (hunk 1 above)
Comment (BenjaminRodenberg) [reply src/cplscheme/BaseCouplingScheme.cpp]: Took care about this in [d3542d4](https://github.com/precice/precice/pull/1887/commits/d3542d4e7596c2bec954f86cf93d9ca00673eef7).

**Pull request merged**